Wangscape/tilegen/TilesetGenerator.cpp

```diff
@@ -51,42 +50,16 @@ void TilesetGenerator::generate(std::function<void(const sf::Texture&, std::stri
 
 void TilesetGenerator::generateClique(const Options::Clique& clique, sf::RenderTexture& image, std::string filename)
 {
-    std::vector<TerrainID> corner_terrains(static_cast<size_t>(CORNERS), clique[0]);
-    std::vector<size_t> corner_clique_indices(static_cast<size_t>(CORNERS), 0);
-    bool stop = false;
-    while (!stop)
+    CornersGenerator cp(clique, static_cast<size_t>(CORNERS));
+    for (auto it = cp.cbegin(); it != cp.cend(); ++it)
     {
-        size_t x = 0; size_t y = 0;
-        for (size_t i = 0; i < corner_clique_indices.size(); i++)
-        {
-            size_t& z = ((i + 1) % 2) ? x : y;
-            z *= clique.size();
-            z += corner_clique_indices[i];
-        }
-        TileGenerator::generate(image, x, y, corner_terrains, images, options, *mTilePartitioner.get());
-        mo.addTile(corner_terrains, filename, x*options.tileFormat.resolution, y*options.tileFormat.resolution);
-        stop = true;
-        auto zip_begin = boost::make_zip_iterator(boost::make_tuple(corner_terrains.begin(),
-                                                                    corner_clique_indices.begin()));
-        auto zip_end = boost::make_zip_iterator(boost::make_tuple(corner_terrains.end(),
-                                                                  corner_clique_indices.end()));
-        for (auto it : boost::make_iterator_range(zip_begin,zip_end))
-        {
-            size_t& i = it.tail.head;
-            TerrainID& t = it.head;
-            i++;
-            if (i >= clique.size())
-            {
-                i = 0;
-                t = clique[i];
-            }
-            else
-            {
-                stop = false;
-                t = clique[i];
-                break;
-            }
-        }
+        const auto& corner_terrains = *it;
+        std::pair<size_t, size_t> tp = it.coordinates_2d();
+        TileGenerator::generate(image, tp.first, tp.second, corner_terrains,
+                                images, options, *mTilePartitioner.get());
+        mo.addTile(corner_terrains, filename,
```

Comment:
The `tiles` field of `MetaOutput` describes the content of a tile (the four corner terrains) and its location (the tileset and offset within the tileset).

The tileset is identified by filename; the only alternatives are the filename with the extension removed, and the list of terrains in the tileset. I prefer the full filename, as it's more explicit.